### PR TITLE
Support GCE via ForwardingRule (LoadBalancer)

### PR DESCRIPTION
**Pull request by justinsb**
Also lots of GCE cleanup

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2176)

### CODE REVIEW

**Comment by justinsb** [comment]
@geojaz did you say you had tried this one?

**Comment by justinsb** [comment]
(I don't think it's necessary to try a PR to give it an LGTM, but if you've tried it and it worked I'd love to get it merged!)

File: pkg/model/gcemodel/api_loadbalancer.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+		// Skipping API LB creation; not requested in Spec
+		return nil
+	}
+
+	switch lbSpec.Type {
+	case kops.LoadBalancerTypePublic:
+	// OK
+
+	case kops.LoadBalancerTypeInternal:
+		return fmt.Errorf("Internal LoadBalancers are not yet supported by kops on GCE")
```
Comment (kris-nova) [inline pkg/model/gcemodel/api_loadbalancer.go]: `s/Internal/internal`

File: pkg/model/gcemodel/api_loadbalancer.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	}
+	// TODO: Health check
+	c.AddTask(forwardingRule)
+
+	{
+		// Ensure the IP address is included in our certificate
+		// TODO: I don't love this technique for finding the task by name & modifying it
+		masterKeypairTask, found := c.Tasks["Keypair/master"]
+		if !found {
+			return fmt.Errorf("Keypair/master task not found")
```
Comment (kris-nova) [inline pkg/model/gcemodel/api_loadbalancer.go]: More capital->lower

(Honestly I really don't care, but I just know we have tried to stick to it in the past)

File: upup/pkg/fi/cloudup/apply_cluster.go [hunk 3]
```diff
@@ -290,13 +290,13 @@ func (c *ApplyClusterCmd) Run() error {
...
 
 			l.AddTypes(map[string]interface{}{
-				"persistentDisk":       &gcetasks.PersistentDisk{},
-				"instance":             &gcetasks.Instance{},
-				"instanceTemplate":     &gcetasks.InstanceTemplate{},
-				"network":              &gcetasks.Network{},
-				"managedInstanceGroup": &gcetasks.ManagedInstanceGroup{},
-				"firewallRule":         &gcetasks.FirewallRule{},
-				"ipAddress":            &gcetasks.IPAddress{},
+				"Disk":                 &gcetasks.Disk{},
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/apply_cluster.go]: I don't think I can express how excited I am to see more tasks being added to GCE. This is so great.

File: upup/pkg/fi/cloudup/gce/gce_url.go [hunk 4]
```diff
@@ -22,21 +22,30 @@ import (
...
+	}
+	url := "https://www.googleapis.com/compute/" + version + "/"
 	if u.Project != "" {
 		url += "projects/" + u.Project + "/"
 	}
 	if u.Global {
 		url += "global/"
 	}
+	if u.Region != "" {
+		url += "regoins/" + u.Region + "/"
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/gce/gce_url.go]: typo?

File: upup/pkg/fi/cloudup/gcetasks/targetpool.go [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+		op, err := t.Cloud.Compute.TargetPools.Insert(t.Cloud.Project, t.Cloud.Region, o).Do()
+		if err != nil {
+			return fmt.Errorf("error creating TargetPool %q: %v", name, err)
+		}
+
+		if err := t.Cloud.WaitForOp(op); err != nil {
+			return fmt.Errorf("error creating TargetPool: %v", err)
+		}
+	} else {
+		return fmt.Errorf("Cannot apply changes to TargetPool: %v", changes)
```
Comment (kris-nova) [inline upup/pkg/fi/cloudup/gcetasks/targetpool.go]: capital again, but you know I don't care just calling it out

File: upup/pkg/fi/cloudup/gce/gce_url.go (hunk 4 above)
Comment (justinsb) [inline upup/pkg/fi/cloudup/gce/gce_url.go]: Wow - thanks!

**Comment by justinsb** [comment]
Thanks - and it's easy to fix the capitalization nits at the start, so call 'em out when you spot em :-)

I agree about list & delete being separate.  We're getting there, in that list can be used both for a `toolbox dump` and for a `delete`, but we can probably do a refactor to make that cleaner still.

**Comment by kris-nova** [comment]
Also tests one day please!

**Pull request merged**